clients/roscpp/include/ros/timer_manager.h

```diff
@@ -428,11 +428,36 @@ void TimerManager<T, D, E>::setPeriod(int32_t handle, const D& period)
   {
     return;
   }
+  
+  if(reset)
+  {
+    info->next_expected = T::now() + period;
```

Comment:
Changing `info->next_expected` should only happen inside the locked mutex scope. Therefore the whole logic from line 432 to 456 should go after line 459.